### PR TITLE
Changes to access control misconfigurations vi uid root

### PR DESCRIPTION
**Pull request by Jjk422**
Separates access control code and user account code from access control modules.
Stops duplicate code in access controls, other then the code that specifies which files are to be changed, and to which user.

Scenario file 'access_control_misconfigurations_uid_vi_root.xml'
Username = 'user'
Password = 'password'

Should be easier to do other uid access control modules, however some code may need to be changed to accommodate changes in pr #68.

### CODE REVIEW

**Comment by thomashaw** [comment]
Looking at this now, cheers :)

**Comment by Jjk422** [comment]
Thanks.
As always if I've made any mistakes, tell me and I'll correct them as soon as I can.

**Comment by thomashaw** [comment]
OK I can see where you're coming from with this, just not 100% sure it's necessary.

Puppet already has the functionality for changing file owners / modes within the default resource tags so this abstraction doesn't seem to add much other than the notice() statement.

file { '/bin/vim'
  mode => x
  owner => x
}

  or

file { ['/bin/vim', '/etc/alternatives/vi', '/usr/bin/vim.tiny']:
 mode => '4777',
 user => 'root'
}


If you want to change a file mode / owner you can just use the standard file resource statements. You can also do multiple at the same time by using an array as above.

If I'm missing the point and there is a good reason for doing it this way let me know and I'll have another proper look. 

Cheers,
Tom

**Comment by Jjk422** [comment]
The original point of doing the modules like this was to ensure that there was one piece of code that does the job for groups of access control modules, therefore any changes that are needed don't need to be copied to all modules individually.
In other words the plan was to have a uid master module, a gid master module and other access control modules that could be used and called easily with data being passed to them so no duplicate code occurred.

This is trivial for uid, but things like gid and more complex access controls may require more lines of code and holding it all in one place, in my opinion is a useful thing to do.

Plus, it makes it more modular, so in my opinion it makes it more maintainable.

I do see what you mean though, we could do it with the file class in each module, although then we won't be able to pass facter/user entered variables/files into it (if each access control vulnerability is defined in the modules themselves) unless we created a separate module for that task.

Thinking in the long run I think a master class for each access control type may be a good thing, but as I wrote the code, I may be slightly biased.

In other words, if you think that it would be better without master modules and with it all defined in individual modules I will change the code to reflect that, although I do think that it may come back to bite us when we get facter implemented properly.

p.s. Sorry for the very long reply.
p.p.s Hopefully you understand the points I am trying to make.

**Comment by thomashaw** [comment]
Good morning, 

I would argue that this makes the program as a whole less 'modular' (i.e. the units of functionality are less independant, separate, interchangable) as each module is now coupled to the 'master' module.

It makes sense for more complex use cases sure, but for the sake of changing access permissions there is already build in functionality that is considered standard.  There is also a 'group' attribute built into puppet to deal with gid here:
https://docs.puppet.com/puppet/latest/reference/type.html#file-attribute-group

There is no real gain in functionality or reduction in code size/readability from what puppet already provides.

Adding another layer of abstraction adds unnecessary complexity and makes our program difficult to follow. (P.S. I am aware that I may have suggested you do this initially but after more investigation it seems that there is a standard way of achieving this).

> "Thinking in the long run I think a master class for each access control type may be a good thing, but as I wrote the code, I may be slightly biased."

I would advise not getting an emotional attachment to code you've written. I've been in your spot before receiving a code review I disagreed with and it's not pleasant. Here's a short article that makes my point for me: http://www.codereadability.com/emotional-attachment-to-code/

Cheers,
Tom

**Comment by Jjk422** [comment]
Thanks for the reply,

After reading your reply and the article links, I more clearly see what you mean. Therefore as "Emotional attachment to code leads to bad decision making". I will "Recognize when it influences [my] decisions and re-examine the choice you've made" (Couldn't resist). 

To be honest another level of abstraction does seem useless now I'm thinking about it, so I am going to redo the modules here either in this pull request or another (depending on the amount of changes needed).

p.s. Thanks for finding that last article, it's an interesting read and puts my techniques of reinventing other people's code into the box of bad ideas.
